test/test_tensorflow.py

```diff
@@ -383,6 +383,118 @@ def test_horovod_allgather(self):
                             tf.equal(tf.cast(rank_tensor, tf.int32), value))),
                         "hvd.allgather produces incorrect gathered tensor")
 
+    def test_horovod_allgather_fused(self):
+        """Test that the allgather correctly gathers 1D, 2D, 3D tensors
+        with Tensor Fusion."""
+        hvd.init()
+        rank = hvd.rank()
+        size = hvd.size()
+
+        with self.test_session(config=self.config) as session:
+            dtypes = [tf.uint8, tf.int8, tf.uint16, tf.int16,
+                      tf.int32, tf.int64, tf.float16, tf.float32,
+                      tf.float64, tf.bool]
+            dims = [1, 2, 3]
+            tests = []
+            tensors_to_gather = []
+            for dtype, dim in itertools.product(dtypes, dims):
+                tensor = tf.ones([17] * dim) * rank
+                if dtype == tf.bool:
+                    tensor = tensor % 2
+                tensor = tf.cast(tensor, dtype=dtype)
+                gathered = hvd.allgather(tensor)
+                tensors_to_gather.append(gathered)
+
+                for i in range(size):
+                    rank_tensor = tf.slice(gathered,
+                                           [i * 17] + [0] * (dim - 1),
+                                           [17] + [-1] * (dim - 1))
+                    # self.assertEqual(list(rank_tensor.shape), [17] * dim)
+                    # tf.equal() does not support tf.uint16 as of TensorFlow 1.2,
+                    # so need to cast rank_tensor to tf.int32.
+                    if dtype != tf.bool:
+                        value = i
+                    else:
+                        value = i % 2
+
+                    tests.append(
+                        tf.reduce_all(
+                            tf.equal(tf.cast(rank_tensor, tf.int32), value)))
+
+            gathered_tensors = session.run(tensors_to_gather)
+            gathered_tensors_it = iter(gathered_tensors)
+            for dtype, dim in itertools.product(dtypes, dims):
+                tensor_shape, expected_shape = list(next(gathered_tensors_it).shape), [17 * size] + [17] * (dim - 1)
```

Comment:
You can use `tf.shape()` and make this condition a TensorFlow Op, that you could add to the list of ops in the loop above.  That way, here you could just do `self.assertTrue(session.run(tf.reduce_all(shape_tests)), ...)`